Sources/TuistLoader/Mappers/ProjectDescriptionHelpersLocationMapper.swift

```diff
@@ -0,0 +1,127 @@
+
+        return rootDirectory
+            .appending(component: Constants.tuistDirectoryName)
+            .appending(component: Constants.helpersDirectoryName)
+            .appending(component: Constants.DerivedDirectory.name)
+    }()
+
+    public init(
+        config: TuistCore.Config,
+        rootDirectoryLocator: RootDirectoryLocating = RootDirectoryLocator()
+    ) {
+        self.config = config
+        self.rootDirectoryLocator = rootDirectoryLocator
+    }
+
+    // MARK: - ProjectMapping
+
+    public func map(project: TuistCore.Project) throws -> (TuistCore.Project, [SideEffectDescriptor]) {
+        guard let helpersLocation = config.helpersLocation else {
+            return (project, [])
+        }
+
+        logger.notice("Using \(helpersLocation) for ProjectDescriptionHelpers")
+
+        switch helpersLocation {
+        case let .local(path):
+            return (project, try collectLocalHelpers(at: path))
+        case let .gitWithBranch(url, gitId),
+             let .gitWithHash(url, gitId),
+             let .gitWithTag(url, gitId):
+            return (project, try collectGitHelpers(at: url, gitId: gitId))
+        }
+    }
+
+    /// Collects files from a local path and copies them to the `ProjectDescriptionHelpers` location.
+    private func collectLocalHelpers(at path: String) throws -> [SideEffectDescriptor] {
+        guard let derivedHelpersDirectory = derivedHelpersDirectory else {
+            logger.error("Unable to locate derived helpers directory.")
+            return []
+        }
+
+        let localHelpersPath = AbsolutePath(path)
+
+        let sources = try FileHandler.shared
+            .contentsOfDirectory(localHelpersPath)
+            .filter { $0.extension == "swift" }
+
+        let sourceContents = try sources.map { try FileHandler.shared.readFile($0) }
+        let sourceNames = sources.map(\.basename)
+        let createHelpersDirectory = DirectoryDescriptor(path: derivedHelpersDirectory, state: .present)
+
+        let createFiles = zip(sourceNames, sourceContents)
+            .map { FileDescriptor(path: AbsolutePath(derivedHelpersDirectory, $0), contents: $1, state: .present) }
+            .map(SideEffectDescriptor.file)
+
+        return [.directory(createHelpersDirectory)] + createFiles
+    }
+
+    private func collectGitHelpers(at url: String, gitId: String) throws -> [SideEffectDescriptor] {
+        guard let derivedHelpersDirectory = derivedHelpersDirectory else {
+            logger.error("Unable to locate derived helpers directory.")
+            return []
+        }
+
+        let helpersGitDirectory = derivedHelpersDirectory.appending(.init(".git"))
+
+        let deleteHelpersDirectory = DirectoryDescriptor(path: derivedHelpersDirectory, state: .absent)
+        let createHelpersDirectory = DirectoryDescriptor(path: derivedHelpersDirectory, state: .present)
+        let gitClone = CommandDescriptor(command: ["git", "clone", url, derivedHelpersDirectory.pathString])
```

Comment:
It is a little bit unfortunate that this clones the repository every time user generates a project. I think that maybe we can save the latest `gitId` and `url` into `Derived` directory and if its contents match to the one supplied to this function, we would not clone it at all? What do you think?

Reply:
For storing the previously pulled git repo how should we serialize this info? Should we just encode the HelpersLocation struct to a file in Derived and then when performing the mapping decode it and if they match just skip downloading?

Reply:
yeah, I think that's a good approach 👍